pkg/services/store/entity/tests/server_integration_test.go

```diff
@@ -109,6 +111,138 @@ func requireVersionMatch(t *testing.T, obj *entity.Entity, m objectVersionMatche
 	require.True(t, len(mismatches) == 0, mismatches)
 }
 
+func TestIntegrationWatch(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	testCtx := createTestContext(t)
+
+	// Update env with entity_api db config
+	dbType := os.Getenv("GRAFANA_TEST_DB")
+	err := addUnifiedStorageConfig(t, testCtx, dbType)
+	require.NoError(t, err)
+
+	t.Run("test should not receive events for keys we are not watching", func(t *testing.T) {
+		group := "test.grafana.app"
+		resource := "jsonobjs"
+		namespace := "default"
+		body := []byte("{\"name\":\"John\"}")
+		name := "test1"
+		key := "/" + group + "/" + resource + "/" + namespace + "/" + name
+		otherKey := "/" + group + "/" + resource + "/" + namespace + "/" + "otherName"
+
+		// create watch client and timeout after 5 seconds of waiting for an event
+		ctx := testCtx.ctx
+		ctx, _ = context.WithTimeout(ctx, time.Second*5)
+		watchClient := newWatchClient(t, ctx, testCtx.client, otherKey)
+
+		// create entity
+		createReq := &entity.CreateEntityRequest{
+			Entity: &entity.Entity{
+				Key:       key,
+				Group:     group,
+				Resource:  resource,
+				Namespace: namespace,
+				Name:      name,
+				Body:      body,
+				Message:   "first entity!",
+			},
+		}
+		_, err = testCtx.client.Create(ctx, createReq)
+		require.NoError(t, err)
+
+		//delete entity
+		_, err = testCtx.client.Delete(ctx, &entity.DeleteEntityRequest{Key: key})
+		require.NoError(t, err)
+
+		// watch client should receive nothing and timeout after 5 seconds
+		_, err = watchClient.Recv()
+		require.Error(t, err)
+		require.ErrorContainsf(t, err, "context deadline exceeded", err.Error())
+	})
+
+	t.Run("watch will receive events for create, update, and delete", func(t *testing.T) {
+		if testing.Short() {
+			t.Skip("skipping integration test")
+		}
+
+		group := "test.grafana.app"
+		resource := "jsonobjs"
+		namespace := "default"
+		body := []byte("{\"name\":\"John\"}")
+		name := "test2"
+		key := "/" + group + "/" + resource + "/" + namespace + "/" + name
+
+		// create watch client and listen for events
+		watchClient := newWatchClient(t, testCtx.ctx, testCtx.client, key)
+
+		// create entity
+		createReq := &entity.CreateEntityRequest{
+			Entity: &entity.Entity{
+				Key:       key,
+				Group:     group,
+				Resource:  resource,
+				Namespace: namespace,
+				Name:      name,
+				Body:      body,
+				Message:   "first entity!",
+			},
+		}
+		_, err = testCtx.client.Create(testCtx.ctx, createReq)
+		require.NoError(t, err)
+
+		// watch client receives create
+		res, err := watchClient.Recv()
```

Comment:
@DanCech This keeps failing because it receives an event for an entity from the previous table test (key `test1`). The watch client didn't subscribe to this key in either test. I'm struggling to see what the problem is here.

The previous test tests for this exact issue (only get watch events for keys you've subbed to), and it passes 🤷‍♂️